plat/common/lcpu.c

```diff
@@ -30,14 +37,141 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <uk/essentials.h>
+#include <uk/arch/atomic.h>
 #include <uk/plat/lcpu.h>
+#include <uk/plat/irq.h>
+#include <uk/plat/time.h>
 #include <uk/plat/common/cpu.h>
+#include <uk/plat/common/lcpu.h>
 #include <uk/plat/common/_time.h>
+#include <uk/list.h>
+#include <uk/print.h>
 
-void ukplat_lcpu_halt(void)
+#include <limits.h>
+#include <errno.h>
+
+/**
+ * Array of LCPUs, one for every CPU in the system.
+ *
+ * TODO: Preferably have a more flexible solution that does not waste memory
+ * for non-present CPUs and does not force us to configure the maximum number
+ * of CPUs beforehand.
+ */
+struct lcpu lcpus[CONFIG_UKPLAT_LCPU_MAXCOUNT] __align(CACHE_LINE_SIZE);
+
+#ifdef CONFIG_HAVE_SMP
+/**
+ * Number of allocated logical CPUs in the system.
+ * Must be [1, CONFIG_UKPLAT_LCPU_MAXCOUNT) after boot
+ */
+static __u32 lcpu_count = 1;
+
+struct lcpu *lcpu_alloc(__lcpuid id)
+{
+	struct lcpu *lcpu;
+
+	if (lcpu_count == CONFIG_UKPLAT_LCPU_MAXCOUNT)
+		return NULL;
+
+	lcpu = &lcpus[lcpu_count];
+	lcpu->state = LCPU_STATE_OFFLINE;
+	lcpu->id    = id;
+	lcpu->idx   = lcpu_count;
+
+	lcpu_count++;
+
+	ukarch_spin_init(&lcpu->lock);
+
+	return lcpu;
+}
+#else
+#define lcpu_count	(1)
+#endif /* !CONFIG_HAVE_SMP */
+
+struct lcpu *lcpu_get(__lcpuidx idx)
+{
+	UK_ASSERT(idx < lcpu_count);
+
+	return &lcpus[idx];
+}
+
+struct lcpu *lcpu_get_current(void)
+{
+	return lcpu_get(ukplat_lcpu_idx());
+}
+
+static void __noreturn lcpu_halt(struct lcpu *this_cpu, int error_code)
 {
 	ukplat_lcpu_disable_irq();
-	halt();
+
+	this_cpu->state = LCPU_STATE_HALTED;
+	this_cpu->error_code = error_code;
+
+	while (1) {
+		/* Although we should not be able to recover via regular
+		 * interrupts, we might receive NMIs so loop to be safe.
+		 */
+		halt();
+	}
+}
+
+int lcpu_init(struct lcpu *this_lcpu)
+{
+	int rc;
+
+	UK_ASSERT(ukplat_lcpu_irqs_disabled());
+	UK_ASSERT(this_lcpu->state == LCPU_STATE_INIT);
+
+	UK_INIT_LIST_HEAD(&this_lcpu->fnlist);
```

Comment:
The bootcore has to run `lcpu_init()` and `lcpu_mp_init()`.

But I am thinking of a plan for getting rid of `lcpu_init()` (at least for the secondary CPUs) by setting up some per-CPU data structures like the GDT and TSS in `lcpu_arch_mp_init()` when discovering the CPUs. Then we only have to activate these when booting a secondary core in the ASM trampoline, which should work. That makes it possible to simply run `main()` functions with prepared stacks as you suggested.

Reply:
One question:

If you directly run `main`, you enable interrupts in the ASM trampoline?

Reply:
I think it's safe to assume that everyone expects interrupts to be already enabled when they enter `main()`.

Reply:
> The bootcore has to run lcpu_init() and lcpu_mp_init().

Right, I thought the platform was only expected to call `lcpu_mp_init()` 👍🏼

Reply:
> The bootcore has to run lcpu_init() and lcpu_mp_init().

But I think that won't work as `lcpu_init()` asserts that cpu->state is `LCPU_STATE_INIT`, yet that is only set by `ukplat_lcpu_start()` on secondary cores.

Reply:
Changed the code so that initialization for the boot core is done in `lcpu_init()` and not `lcpu_mp_init()` so it is much earlier in the right state. Also adapted the `UK_ASSERT` check for `INIT` state as you suggested 👍🏻